### PR TITLE
Assorted Typing Fixes

### PR DESCRIPTION
**Pull request by askvortsov1**
Moving one step closer to having all the hundreds of CI jobs on this monorepo be green.

### CODE REVIEW

**Review by SychO9 — APPROVED** [review]
Amazing to see this!

Wouldn't have been fun across multiple repositories >.>

Looks like the diff shows dist files despite `.gitattributes` files instructing against, it seems like only a root level `.gitattributes` is taken into consideration so we'll probably want to add one.

Other than that, all looks good to me! Thanks for the well-split commits.

File: extensions/akismet/js/src/forum/index.ts [hunk 1]
```diff
@@ -5,14 +5,18 @@ import PostControls from 'flarum/forum/utils/PostControls';
 import CommentPost from 'flarum/forum/components/CommentPost';
 import ItemList from 'flarum/common/utils/ItemList';
 import Post from 'flarum/common/models/Post';
+import Mithril from 'mithril';
```
Comment (davwheat) [inline extensions/akismet/js/src/forum/index.ts]:
```suggestion
import type Mithril from 'mithril';
```

File: extensions/flags/js/src/forum/index.ts [hunk 2]
```diff
@@ -26,5 +26,6 @@ app.initializers.add('flarum-flags', () => {
 // Expose compat API
 import flagsCompat from './compat';
 import { compat } from '@flarum/core/forum';
+import Post from 'flarum/common/models/Post';
```
Comment (davwheat) [inline extensions/flags/js/src/forum/index.ts:29]: I think this should be at the top of the file instead, for clarity.

File: extensions/flags/js/src/forum/models/Flag.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import Model from 'flarum/common/Model';
+import Post from 'flarum/common/models/Post';
```
Comment (davwheat) [inline extensions/flags/js/src/forum/models/Flag.ts]:
```suggestion
import type Post from 'flarum/common/models/Post';
```

File: extensions/flags/js/src/forum/models/Flag.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import User from 'flarum/common/models/User';
```
Comment (davwheat) [inline extensions/flags/js/src/forum/models/Flag.ts]:
```suggestion
import type User from 'flarum/common/models/User';
```

File: extensions/pusher/js/src/forum/index.ts [hunk 5]
```diff
@@ -6,7 +6,8 @@ import DiscussionPage from 'flarum/forum/components/DiscussionPage';
 import IndexPage from 'flarum/forum/components/IndexPage';
 import Button from 'flarum/common/components/Button';
 import ItemList from 'flarum/common/utils/ItemList';
-import { Children, VnodeDOM } from 'mithril';
+import { Children } from 'mithril';
```
Comment (davwheat) [inline extensions/pusher/js/src/forum/index.ts]:
```suggestion
import type { Children } from 'mithril';
```

File: extensions/pusher/js/src/forum/index.ts [hunk 6]
```diff
@@ -6,7 +6,8 @@ import DiscussionPage from 'flarum/forum/components/DiscussionPage';
...
+import Tag from 'flarum/tags/common/models/Tag';
```
Comment (davwheat) [inline extensions/pusher/js/src/forum/index.ts]:
```suggestion
import type Tag from 'flarum/tags/common/models/Tag';
```

File: extensions/tags/js/dist-typings/@types/shims.d.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import Tag from "../common/models/Tag";
```
Comment (davwheat) [inline extensions/tags/js/dist-typings/@types/shims.d.ts]:
```suggestion
import type Tag from "../common/models/Tag";
```

File: extensions/tags/js/dist-typings/@types/shims.d.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import TagListState from "../forum/states/TagListState";
```
Comment (davwheat) [inline extensions/tags/js/dist-typings/@types/shims.d.ts]:
```suggestion
import type TagListState from "../forum/states/TagListState";
```

File: extensions/tags/js/src/@types/shims.d.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import Tag from "../common/models/Tag";
```
Comment (davwheat) [inline extensions/tags/js/src/@types/shims.d.ts]:
```suggestion
import type Tag from "../common/models/Tag";
```

File: extensions/tags/js/src/@types/shims.d.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import TagListState from "../forum/states/TagListState";
```
Comment (davwheat) [inline extensions/tags/js/src/@types/shims.d.ts]:
```suggestion
import type TagListState from "../forum/states/TagListState";
```

File: extensions/tags/js/src/common/models/Tag.ts [hunk 11]
```diff
@@ -0,0 +1,71 @@
+import computed from 'flarum/common/utils/computed';
+import Model from 'flarum/common/Model';
+import Discussion from 'flarum/common/models/Discussion';
```
Comment (davwheat) [inline extensions/tags/js/src/common/models/Tag.ts]:
```suggestion
import type Discussion from 'flarum/common/models/Discussion';
```

File: extensions/tags/js/src/forum/states/TagListState.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
+import app from "flarum/forum/app";
+import Tag from "../../common/models/Tag";
```
Comment (davwheat) [inline extensions/tags/js/src/forum/states/TagListState.ts]:
```suggestion
import type Tag from "../../common/models/Tag";
```

File: framework/core/js/dist-typings/admin/AdminApplication.d.ts [hunk 13]
```diff
@@ -1,3 +1,4 @@
+import { AdminRoutes } from './routes';
```
Comment (davwheat) [inline framework/core/js/dist-typings/admin/AdminApplication.d.ts:1]:
```suggestion
import type { AdminRoutes } from './routes';
```

File: framework/core/js/dist-typings/admin/components/HeaderPrimary.d.ts [hunk 14]
```diff
@@ -4,6 +4,7 @@
  */
 export default class HeaderPrimary extends Component<import("../../common/Component").ComponentAttrs, undefined> {
     constructor();
+    view(): JSX.Element;
```
Comment (davwheat) [inline framework/core/js/dist-typings/admin/components/HeaderPrimary.d.ts:7]: All of these should really be `Mithril.Children` instead (return type manually defined in the method declaration).

File: framework/core/js/src/admin/components/DashboardPage.tsx [hunk 15]
```diff
@@ -3,6 +3,7 @@ import StatusWidget from './StatusWidget';
 import ExtensionsWidget from './ExtensionsWidget';
 import ItemList from '../../common/utils/ItemList';
 import AdminPage from './AdminPage';
+import { Children } from 'mithril';
```
Comment (davwheat) [inline framework/core/js/src/admin/components/DashboardPage.tsx]:
```suggestion
import type { Children } from 'mithril';
```

File: framework/core/js/src/admin/components/DashboardWidget.tsx [hunk 16]
```diff
@@ -0,0 +1,18 @@
+import { Children, Vnode } from 'mithril';
```
Comment (davwheat) [inline framework/core/js/src/admin/components/DashboardWidget.tsx]:
```suggestion
import type { Children, Vnode } from 'mithril';
```

File: framework/core/js/src/forum/ForumApplication.ts [hunk 17]
```diff
@@ -23,20 +23,22 @@ import isSafariMobile from './utils/isSafariMobile';
 import type Notification from './components/Notification';
 import type Post from './components/Post';
 import Discussion from '../common/models/Discussion';
+import NotificationModel from '../common/models/Notification';
```
Comment (davwheat) [inline framework/core/js/src/forum/ForumApplication.ts]:
```suggestion
import type NotificationModel from '../common/models/Notification';
```

File: framework/core/js/src/forum/ForumApplication.ts [hunk 18]
```diff
@@ -23,20 +23,22 @@ import isSafariMobile from './utils/isSafariMobile';
...
+import PostModel from '../common/models/Post';
```
Comment (davwheat) [inline framework/core/js/src/forum/ForumApplication.ts]:
```suggestion
import type PostModel from '../common/models/Post';
```

File: framework/core/js/src/forum/ForumApplication.ts [hunk 19]
```diff
@@ -23,20 +23,22 @@ import isSafariMobile from './utils/isSafariMobile';
 import type Notification from './components/Notification';
 import type Post from './components/Post';
 import Discussion from '../common/models/Discussion';
```
Comment (davwheat) [inline framework/core/js/src/forum/ForumApplication.ts]: I know this wasn't part of your changes, but it shouldn't hurt.

```suggestion
import type Discussion from '../common/models/Discussion';
```

File: framework/core/js/src/forum/components/DiscussionListItem.tsx [hunk 20]
```diff
@@ -17,26 +17,31 @@ import classList from '../../common/utils/classList';
 import DiscussionPage from './DiscussionPage';
 import escapeRegExp from '../../common/utils/escapeRegExp';
 import Tooltip from '../../common/components/Tooltip';
+import Discussion from '../../common/models/Discussion';
```
Comment (davwheat) [inline framework/core/js/src/forum/components/DiscussionListItem.tsx]:
```suggestion
import type Discussion from '../../common/models/Discussion';
```

File: framework/core/js/src/forum/components/DiscussionListItem.tsx [hunk 21]
```diff
@@ -17,26 +17,31 @@ import classList from '../../common/utils/classList';
...
+import Mithril from 'mithril';
```
Comment (davwheat) [inline framework/core/js/src/forum/components/DiscussionListItem.tsx]:
```suggestion
import type Mithril from 'mithril';
```

File: framework/core/js/src/forum/components/DiscussionListItem.tsx [hunk 22]
```diff
@@ -17,26 +17,31 @@ import classList from '../../common/utils/classList';
...
+import { DiscussionListParams } from '../states/DiscussionListState';
```
Comment (davwheat) [inline framework/core/js/src/forum/components/DiscussionListItem.tsx]:
```suggestion
import type { DiscussionListParams } from '../states/DiscussionListState';
```

File: framework/core/js/src/forum/components/DiscussionRenamedNotification.tsx [hunk 23]
```diff
@@ -1,13 +1,14 @@
+import Discussion from '../../common/models/Discussion';
```
Comment (davwheat) [inline framework/core/js/src/forum/components/DiscussionRenamedNotification.tsx]:
```suggestion
import type Discussion from '../../common/models/Discussion';
```

File: framework/core/js/src/forum/components/DiscussionRenamedNotification.tsx [hunk 24]
```diff
@@ -16,11 +17,20 @@ export default class DiscussionRenamedNotification extends Notification {
...
-    return app.route.discussion(notification.subject(), notification.content().postNumber);
+    return app.route.discussion(discussion as Discussion, notification.content<DiscussionRenamedContent>().postNumber);
   }
 
   content() {
     return app.translator.trans('core.forum.notifications.discussion_renamed_text', { user: this.attrs.notification.fromUser() });
   }
+
+  excerpt() {
+    return '';
```
Comment (davwheat) [inline framework/core/js/src/forum/components/DiscussionRenamedNotification.tsx]: Does it make more sense to return `null` instead?

File: framework/core/js/src/forum/components/Notification.tsx [hunk 25]
```diff
@@ -1,35 +1,37 @@
 import app from '../../forum/app';
-import Component from '../../common/Component';
+import NotificationModel from '../../common/models/Notification';
```
Comment (davwheat) [inline framework/core/js/src/forum/components/Notification.tsx]:
```suggestion
import type NotificationModel from '../../common/models/Notification';
```

File: framework/core/js/src/forum/components/Notification.tsx [hunk 26]
```diff
@@ -1,35 +1,37 @@
...
+import Component, { ComponentAttrs } from '../../common/Component';
 import avatar from '../../common/helpers/avatar';
 import icon from '../../common/helpers/icon';
 import humanTime from '../../common/helpers/humanTime';
 import Button from '../../common/components/Button';
 import Link from '../../common/components/Link';
 import classList from '../../common/utils/classList';
+import Mithril from 'mithril';
```
Comment (davwheat) [inline framework/core/js/src/forum/components/Notification.tsx]:
```suggestion
import type Mithril from 'mithril';
```

File: framework/core/js/src/forum/components/Post.tsx [hunk 27]
```diff
@@ -1,53 +1,54 @@
 import app from '../../forum/app';
-import Component from '../../common/Component';
+import Component, { ComponentAttrs } from '../../common/Component';
 import SubtreeRetainer from '../../common/utils/SubtreeRetainer';
 import Dropdown from '../../common/components/Dropdown';
 import PostControls from '../utils/PostControls';
 import listItems from '../../common/helpers/listItems';
 import ItemList from '../../common/utils/ItemList';
+import PostModel from '../../common/models/Post';
```
Comment (davwheat) [inline framework/core/js/src/forum/components/Post.tsx]:
```suggestion
import type PostModel from '../../common/models/Post';
```

File: framework/core/js/src/forum/components/Post.tsx [hunk 28]
```diff
@@ -1,53 +1,54 @@
...
 import LoadingIndicator from '../../common/components/LoadingIndicator';
+import Mithril from 'mithril';
```
Comment (davwheat) [inline framework/core/js/src/forum/components/Post.tsx]:
```suggestion
import type Mithril from 'mithril';
```

File: framework/core/js/src/forum/routes.ts [hunk 29]
```diff
@@ -10,6 +10,15 @@ import Discussion from '../common/models/Discussion';
 import Post from '../common/models/Post';
```
Comment (davwheat) [inline framework/core/js/src/forum/routes.ts]:
```suggestion
import type Post from '../common/models/Post';
```

File: framework/core/js/src/forum/routes.ts [hunk 30]
```diff
@@ -10,6 +10,15 @@ import Discussion from '../common/models/Discussion';
...
 import User from '../common/models/User';
```
Comment (davwheat) [inline framework/core/js/src/forum/routes.ts]:
```suggestion
import type User from '../common/models/User';
```

File: framework/core/js/src/forum/states/PostStreamState.ts [hunk 31]
```diff
@@ -1,55 +1,71 @@
 import app from '../../forum/app';
 import { throttle } from 'throttle-debounce';
 import anchorScroll from '../../common/utils/anchorScroll';
+import Discussion from '../../common/models/Discussion';
```
Comment (davwheat) [inline framework/core/js/src/forum/states/PostStreamState.ts]:
```suggestion
import type Discussion from '../../common/models/Discussion';
```

File: framework/core/js/src/forum/states/PostStreamState.ts [hunk 32]
```diff
@@ -1,55 +1,71 @@
...
+import Post from '../../common/models/Post';
```
Comment (davwheat) [inline framework/core/js/src/forum/states/PostStreamState.ts]:
```suggestion
import type Post from '../../common/models/Post';
```

File: framework/core/js/src/forum/states/PostStreamState.ts [hunk 33]
```diff
@@ -1,55 +1,71 @@
...
-     * The discussion to display the post stream for.
-     *
-     * @type {Discussion}
-     */
-    this.discussion = discussion;
+export default class PostStreamState {
+  /**
+   * @see https://github.com/Microsoft/TypeScript/issues/3841#issuecomment-337560146
+   */
+  ['constructor']: typeof PostStreamState;
```
Comment (davwheat) [inline framework/core/js/src/forum/states/PostStreamState.ts:11]: Hah, clever!

**Review by davwheat — CHANGES_REQUESTED** [review]
Most of these are just using `import type` instead of `import` to make certain that we don't accidentally use the imports for anything except typing.

Another thing to note is that we should *really* be marking view method return types as `Mithril.Children` rather than the automatic `JSX.Element` since that doesn't allow extends/overrides to return things like `null`, strings, numbers or arrays as fragments.

Thank you for your hard work on this 🤗🥰

File: framework/core/js/src/forum/states/PostStreamState.ts (hunk 33 above)
Comment (askvortsov1) [reply framework/core/js/src/forum/states/PostStreamState.ts:11]: All credit goes to the one guy on stack overflow :wink:

**Review by davwheat — APPROVED** [review]

**Pull request merged**